### PR TITLE
Implement payments via Express Checkout Element for Simple Products on the Product Page 

### LINKED ISSUE #8868: Integrate ECE for Payment Processing in the product page
Implement the Stripe Express Checkout Element (ECE) to enable direct payment processing from the product page, bypassing the shopper cart. 

### Acceptance criteria:

- The ECE must successfully process payments directly from the product page.
- The behavior of express checkout buttons on the product page should mirror that of Payment Request Buttons (PRBs).
- Ensure there are no regressions in Payment Request Buttons (PRB) when ECE is disabled.
- There are no regressions when no express checkout payment methods are active.

### PR DESCRIPTION
**Pull request by asumaran**
Fixes #8868 

#### Changes proposed in this Pull Request

Implements payments using the Express Checkout Elements API.
This feature is behind the `_wcpay_feature_stripe_ece` flag.

#### Testing instructions

Note that the changes on this PR are the same as #8914. So, please  make sure to follow the setup instructions before starting to test this PR.

Testing on this PR must focus on the product page for **simple products** only.

> [!NOTE]
> There shouldn't be any problem with testing Google Pay or Apple Pay using your own credit card. **Just make sure the site is in test mode**.

Visit the product page for a simple product then:

**Pay using Google Pay**
- Verify shipping address can be changed.
- Verify the order is created with correct amount.
- Verify line items are present in the payment sheet.
- Verify tax is applied to the totals.
- Test with a US shipping address.
- Test with a US billing address.

**Pay using Apple Pay**
- Verify shipping address can be changed.
- Verify the order is created with correct amount.
- Verify line items are present in the payment sheet.
- Verify tax is applied to the totals.
- Test with a US shipping address.
- Test with a US billing address.

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8937` or branch name `as-8868-ece-product-page-2` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=as-8868-ece-product-page-2) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 5cf5662e2d244623c4989b6f504e389ee3b2ad74
- Build time: 2024-06-18 18:18:57 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Comment by ricardo** [comment]
@asumaran As noted from #8914, the checkout from the product page didn't work for me. I figured out why:

In case the fraud prevention service [is enabled](https://github.com/Automattic/woocommerce-payments/blob/b73ef0649a6b98c68607de329567ef7cb0a856cb/includes/class-wc-payments-account.php#L1997) for the account, I guess we're not passing the token correctly [here](https://github.com/Automattic/woocommerce-payments/blob/bf558f8410585ce7db6b062affa504e926883e79/includes/class-wc-payment-gateway-wcpay.php#L1164), so it throws this error:

> We're not able to process this payment. Please refresh the page and try again.

If I comment out that check I can proceed with the checkout.

**Comment by ricardo** [comment]
Just to add to the above comment. I'm also having problems checking out with a saved card while the fraud prevention service is enabled, so this is either a problem with my local environment or a more general issue.

File: client/express-checkout/event-handlers.js
Comment (lovo-h) [inline client/express-checkout/event-handlers.js]: We may want to `return` here. Otherwise, we will proceed with creating a payment method, etc.

File: client/express-checkout/event-handlers.js
Comment (lovo-h) [inline client/express-checkout/event-handlers.js:62]: This still works but it should be noted that `abortPayment( ... )` returns `void`.

File: client/express-checkout/index.js
Comment (lovo-h) [inline client/express-checkout/index.js:195]: Should we add `product.shippingOptions` to the `WCPayExpressCheckoutParams` interface?
```javascript
// ...
product: {
    needs_shipping: boolean;
    currency: string;
    shippingOptions: {
        id: string;
        label: string;
        detail: string;
        amount: number;
    };
};
// ...
```

File: client/express-checkout/index.js
Comment (lovo-h) [inline client/express-checkout/index.js:250]: Should we `await` for the product to be added to the cart before proceeding? Alternatively, we may create a promise that resolves later.

File: client/express-checkout/index.js
Comment (lovo-h) [inline client/express-checkout/index.js:268]: It looks like we are not using the `async`, can it be removed?

File: client/express-checkout/index.js
Comment (lovo-h) [inline client/express-checkout/index.js:272]: It looks like we are not using the `async`, can it be removed?

File: client/express-checkout/index.js
Comment (lovo-h) [inline client/express-checkout/index.js]: I am not seeing where `paymentAborted` is being used for `express-checkout`. Will it be used, later? Or can it be removed?

File: client/express-checkout/index.js
Comment (lovo-h) [inline client/express-checkout/index.js:409]: Should we also update this one?
```javascript
getExpressCheckoutData( 'product' )?.displayItems,
```

File: includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php
Comment (lovo-h) [inline includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php]: Can we refactor this to the following:
```php
// ...
$data           = $this->express_checkout_button_helper->build_display_items( $should_show_itemized_view );
$data['result'] = 'success';
// ...
```

**Review by lovo-h — CHANGES_REQUESTED** [review]
Code changes are looking good and testing well.

I left a few comments we may want to address before merging.

**Testing Instructions**
* ✅ Pay using Google Pay. All tests passed.
* ✅ Pay using Apple Pay. All tests passed.
* ✅ `npm run test:js` and `npm run test:php`.

File: client/express-checkout/event-handlers.js
Comment (reykjalin) [reply client/express-checkout/event-handlers.js]: Nice catch! Fixed in https://github.com/Automattic/woocommerce-payments/commit/c79db1e8fef0b3a4e2023b826222b2443ff2e33b

File: client/express-checkout/event-handlers.js
Comment (reykjalin) [reply client/express-checkout/event-handlers.js:62]: We don't enforce a specific return type here and I saw this pattern in the PRB implementation so I figured it'd be nice to give the caller the ability to do something with a returned value when the payment is aborted.

Returning `void` just means the caller chose to do nothing 🙂

File: client/express-checkout/index.js
Comment (reykjalin) [reply client/express-checkout/index.js:195]: Yes! I didn't because I wasn't working on the product page, so it'd be great if @asumaran could add the type information.

File: client/express-checkout/index.js
Comment (reykjalin) [reply client/express-checkout/index.js:268]: I actually went the opposite way and used it everywhere for consistency in https://github.com/Automattic/woocommerce-payments/commit/793765648dd7d1bdcf2c193013ed2f75ab31464e based on https://github.com/Automattic/woocommerce-payments/pull/8914#discussion_r1635494753, but I'm fine with removing it too. @ricardo what do you think?

File: client/express-checkout/index.js
Comment (reykjalin) [reply client/express-checkout/index.js:272]: See https://github.com/Automattic/woocommerce-payments/pull/8937#discussion_r1637342212

File: client/express-checkout/index.js
Comment (reykjalin) [reply client/express-checkout/index.js]: I don't know where it's used either. I didn't touch it because I was worried it'd break _something_. It might be used for tracks somewhere? I didn't look into it, but I suspect it's something that was copied out of the PRB implementation.

File: client/express-checkout/index.js
Comment (reykjalin) [reply client/express-checkout/index.js:409]: I'd rather not introduce a bunch of unrelated changes to the PR, so let's keep pre-existing cases as they are and use `getExpressCheckoutData` for new code.

I'm not super tied to that opinion though, happy to make changes where we deem it appropriate.

File: includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php
Comment (reykjalin) [reply includes/express-checkout/class-wc-payments-express-checkout-ajax-handler.php]: Fixed in 000f8ad4270f52e43ea22157b154a641c398361a

File: client/express-checkout/index.js
Comment (asumaran) [reply client/express-checkout/index.js:250]: We can't due to the [1 second limitation](https://docs.stripe.com/js/elements_object/express_checkout_element_click_event#express_checkout_element_on_click-handler-resolve). We have to call the resolve function before we run out of time. If we wait for the ajax request to finish it will throw an error. 

I think it's OK since the payment details are later updated when the payment sheet is displayed anyway.

It would be helpful to test what happens if adding an item to the cart takes too long and the user attempts to submit the payment sheet before it finishes. I'll create an issue to investigate this scenario 👍.

File: client/express-checkout/index.js
Comment (asumaran) [reply client/express-checkout/index.js:195]: done [54411c0](https://github.com/Automattic/woocommerce-payments/pull/8937/commits/54411c01fe272ad544b10d5f5a7f7f5381e7adb5)

File: client/express-checkout/index.js
Comment (asumaran) [reply client/express-checkout/index.js:268]: It feels strange to see the async keyword without an await inside. However, it doesn't hurt to include it for the sake of consistency.

File: client/express-checkout/index.js
Comment (asumaran) [reply client/express-checkout/index.js:250]: Issue created here. I'll work on that later. https://github.com/Automattic/woocommerce-payments/issues/8967

File: client/express-checkout/index.js
Comment (asumaran) [reply client/express-checkout/index.js]: It was copied from the PRB implementation. For PRBs, it's [used](https://github.com/Automattic/woocommerce-payments/blob/develop/client/payment-request/index.js#L477) to reinitialize the button when the variation changes.

I'm going to remove it since we don't use it currently. If necessary, we could reintroduce it when working on [#8871](https://github.com/Automattic/woocommerce-payments/issues/8871) .

Comment (asumaran) [reply client/express-checkout/index.js]: done [766a675](https://github.com/Automattic/woocommerce-payments/pull/8937/commits/766a675a1eb1b0062d2788bc18611da905bb257d)

**Comment by asumaran** [comment]
@ricardo, I was able to see the bug, too.
However, let's handle it separately since I'm not sure if it's fixable for ECE right now, given that the same error happens for PRBs. What do you think, @bborman22?
That being said, shall we go ahead and merge this pull request?

File: client/express-checkout/index.js
Comment (asumaran) [reply client/express-checkout/index.js:250]: Ran some tests in https://github.com/Automattic/woocommerce-payments/issues/8967#issuecomment-2174696547 and can confirm it's failing as expected.   

Since we can't call the `resolve` function after 1 second, if adding product(s) to the cart takes too long then it should fail gracefully at least.

In the scenarios I've tested, no order is created and the user is not charged. While this is not ideal, it is an acceptable outcome considering the severity of the error.

One approach we could take is to set a flag when the request for `/?wc-ajax=wcpay_add_to_cart` is made. Then, if the shipping address or shipping options are changed, we could reject the event, as it wouldn't be necessary to continue doing that if the product hasn't been added to the cart. I'll attempt to implement this while working on issue #8871.

**Comment by bborman22** [comment]
@asumaran @ricardo Just to confirm this is the bug with `checking out with a saved card while the fraud prevention service is enabled`? And both of you are able to reproduce. If it happens on PRBs as well today, it shouldn't block this PR and we should create an issue to track it instead.

**Comment by asumaran** [comment]
@bborman22 I was referring to the bug while the prevention service is enabled for the site. I didn't test with saved cards but I assume we'll have the same results.

**Review by lovo-h — APPROVED** [review]
Code changes LGTM 👍

**Comment by asumaran** [comment]
@bborman22 @ricardo I've created an issue for the issue we have with ECE and the Card Testing Protection Service. https://github.com/Automattic/woocommerce-payments/issues/8981

**Comment by asumaran** [comment]
@bborman22 @ricardo ^ I'm fixing https://github.com/Automattic/woocommerce-payments/issues/8981 in https://github.com/Automattic/woocommerce-payments/pull/8987. The fix was pretty straightforward https://github.com/Automattic/woocommerce-payments/pull/8987/commits/59f8f3303db7a5fce58f287f3fc37b1690158c0c

**Pull request merged**
